Fuzzy Search

ecxyzzy opened the pull request:
## Summary
Implement fuzzy searching with the `websoc-fuzzy-search` module.

## Testing
https://staging-343.antalmanac.com

## To-do
- [x] #345
- [x] Currently, queries manipulate the `RightPane` directly, which means the legacy search form will reflect any autocompleted fields after the user returns from the search view. Not really sure if there's a cleaner way to go about this. (It's fine.)
## Issues
Closes #266.

ecxyzzy committed:
add websoc-fuzzy-search from local

ecxyzzy committed:
implement fuzzy search component

ecxyzzy committed:
add fuzzy search to SearchForm for testing

ecxyzzy committed:
get rid of dropdown arrow

ecxyzzy committed:
prevent snackbar on failed autocomplete

ecxyzzy committed:
properly update dept and GE fields

ChaseC99 commented on src/components/SearchForm/FuzzySearch.js:
```diff
@@ -0,0 +1,119 @@
+    GE_CATEGORY: '🏫', // U+1F3EB :school:
+    DEPARTMENT: '🏢', // U+1F3E2 :office:
+    COURSE: '📚', // U+1F4DA :books:
+    INSTRUCTOR: '🧑', // U+1F9D1 :adult:
```
Let's use the teacher emoji
```suggestion
    INSTRUCTOR: '🧑‍🏫',
```

ecxyzzy replied:
Slight issue with this: the zero-width joiner doesn't actually render on my laptop, so the teacher "emoji" actually displays as two separate emojis, like so:
![image](https://user-images.githubusercontent.com/89349085/163302186-0eafe2c6-0eae-4188-95fe-b46b76c4f7d8.png)

Can you replicate this on your setup?

ChaseC99 replied:
Hmm are you on Windows or Linux? I'm on Mac

ecxyzzy replied:
I'm on Windows. Did some more experimentation, and it looks like Windows 10's emoji font spec only supports ZWJ on gendered person emojis: 👨‍🏫 (man + school -> male teacher) and 👩‍🏫 (woman + school -> female teacher) render as one emoji, but 🧑‍🏫 (person + school) doesn't.

(For what it's worth, Windows 11 does support ZWJ on [more] gender-neutral person emojis. The downside is having to use Windows 11.)

devsdevsdevs replied:
Classic windows

ChaseC99 replied:
Uh oh so we want to use the male or female one? I feel like the male one looks closer to the gender neutral one

ecxyzzy replied:
Yeah unfortunately I think it's either choosing between them or using the person emoji 😕

> I feel like the male one looks closer to the gender neutral one

That's what I was thinking too, might be our best choice for now...

ecxyzzy replied:
After some discussion on Discord Chase suggested to use 🍎 (apple), which sounds like a good enough idea to me, so it has been implemented in 13f7a30.

ChaseC99 commented on src/components/SearchForm/FuzzySearch.js:
```diff
@@ -0,0 +1,119 @@
+                        'instructor',
+                        Object.keys(this.state.results).filter((x) => this.state.results[x].name === ident)
+                    );
+                }
```
Thoughts on using a switch statement for this?
```suggestion
                switch (emoji) {
                    case emojiMap.GE_CATEGORY:
                        updateFormValue(
                            'ge',
                            `GE-${ident[0].split(' ')[2].replace('(', '').replace(')', '').toUpperCase()}`
                        );
                        break;
                    case emojiMap.DEPARTMENT:
                         updateFormValue('deptValue', ident[0]);
                         updateFormValue('deptLabel', ident.join(':'));
                         break;
                    case emojiMap.COURSE:
                        updateFormValue('deptValue', ident[0].split(' ').slice(0, -1).join(' '));
                        updateFormValue('deptLabel', ident.join(':'));
                        updateFormValue('courseNumber', ident[0].split(' ').slice(-1)[0]);
                        break;
                    case emojiMap.INSTRUCTOR:
                        updateFormValue(
                            'instructor',
                            Object.keys(this.state.results).filter((x) => this.state.results[x].name === ident)
                        );
                        break;
                }
```

ChaseC99 reviewed (changes requested):
Overall looking awesome! Focusing on functionality, here are some bugs I found:

- Department field gets updated to the class title when you press the back button on search results 
<img width="797" alt="image" src="https://user-images.githubusercontent.com/29494270/163298373-f43164a6-0b24-4edd-991d-b28015704037.png">

- Multi word parsing still needs a bit of work

| `ics` | `ics `&nbsp; | `ics 3` | `ics 33` |
| --- | --- | --- | --- |
| Looks good, shows ics results. | Once you add a space, nothing gets returned. Should show same results as `ics` | Also shows no results but it should show the "ICS 3: Internet Technologies and their Social Impact" and the 3X series | Looks good |
| <img width="800" alt="image" src="https://user-images.githubusercontent.com/29494270/163298645-f6e8edc8-1c52-46c3-adc1-f08d8f345b06.png"> | <img width="796" alt="image" src="https://user-images.githubusercontent.com/29494270/163298681-18571c68-3848-4ac4-9609-479c423426e0.png"> | <img width="800" alt="image" src="https://user-images.githubusercontent.com/29494270/163298724-deec564f-e851-4a02-8e11-7a8ebb8f9da8.png"> | <img width="797" alt="image" src="https://user-images.githubusercontent.com/29494270/163298755-262986dc-77dc-433e-9908-8e0710b9de45.png"> |

ecxyzzy committed:
bugfixes and refactoring

- Fixed department field getting updated to the course when such a query
was made
- Switched (no pun intended) if-else ifs to switch-cases

ecxyzzy committed:
improve trailing space and course number parsing

ecxyzzy committed:
updates

- Change instructor emoji to 🍎 (apple)
- Shorten slice index notation

ChaseC99 committed:
Hide LegacySearch  (#344)

* Move old search to LegacySearch component, move term to top, move fuzzy inside of form

* Stick Privacy Banner on the bottom

ecxyzzy committed:
update websoc-fuzzy-search to 0.6.3

ChaseC99 committed:
LegacySearch moved under options dropdown (#346)

- LegacySearch is now nicely hidden under a 'Manual Options' button
- Converted SearchForm to functional component

falsidge committed:
Merge branch 'main' into fuzzy-search

* Add contributors to about modal (#342)
* Colorpicker fix for update and closing (#326)

Co-authored-by: Chase Carnaroli <chasecarnaroli@gmail.com>

Voark committed:
Merge branch 'staging-actions' into fuzzy-search

ecxyzzy committed:
Revert upstream branch merges

This reverts the following commits:
* e8ae381 [main -> fuzzy-search]
* be8731b [staging-actions -> fuzzy-search]

ChaseC99 reviewed (approved):
The revert looks good. Let's ship this feature 🎉
